// @sap/cds-runtime/lib/index.js
module.exports = {
  to: {
    /** @type {import('./cds-services/adapter/odata-v4/to'))  )} */
    get odata_v4() {
      return this._odatav4 || (this._odatav4 = require('./cds-services/adapter/odata-v4/to'))
    },

    /** @type {import('./cds-services/adapter/rest/to'))  )} */
    get rest() {
      return this._rest || (this._rest = require('./cds-services/adapter/rest/to'))
    }
  },

  /** @type {import('./common/auth')))} */
  get auth() {
    return this._auth || (this._auth = require('./common/auth'))
  },

  /** @type {import('./cds-services/statements')))} */
  get statements() {
    return this._statements || (this._statements = require('./cds-services/statements'))
  },

  /** @type {import('../package.json').version))} */
  get version() {
    return this._version || (this._version = require('../package.json').version)
  },

  // REVISIT: remove once not needed anymore
  get performanceMeasurement() {
    return this._perf || (this._perf = require('./cds-services/adapter/perf/performanceMeasurement'))
  },

  // REVISIT: remove once not needed anymore
  get passport() {
    return this.auth
  }
}
